### ISSUE TITLE
startup fails if installation folder name includes Chinese chars

### ISSUE THREAD

**Issue by imbear** [neovim/neovim#9267]
NVIM v0.3.2-913-g108566e7b
Build type: RelWithDebInfo
LuaJIT 2.0.5

Win10 home Chinese

I extract nvim to a folder with a Chinese char, so nvim will run failed.

`<d6><d0>` is cp936 code of `中`  

```
@ D:\Neovim中\bin
$ nvim
C:\Users\boom\vimfiles\plugin.vimrc
Vim(source):E484: € D:\Neovim<d6><d0>\share\nvim\runtime\syntax\syntax.vim
C:\Users\boom\vimfiles\face.vimrc
Vim(source):E484: € D:\Neovim<d6><d0>\share\nvim\runtime\syntax\syntax.vim
 ENTER
```

**Comment by justinmk** [comment]
on macOS:

```
mkdir -p Neovim中/foo
touch Neovim中/foo/vimrc
nvim -u Neovim中/foo/vimrc
```

**Comment by erw7** [comment]
I think this is the same problem as #8398. Since environment variables `VIM` and `VIMRUNTIME` contain multibyte characters, I think that `os_getenv` returns the character string of `cp936`. The phenomenon is reproduced also in my environment of Windows 7 Japanese.

I also found a problem with environment variables such as `os_getenvname_at_index` on Windows. @justinmk Are you planning on implementing #7920? I do not mind taking over #7920.

**Comment by justinmk** [comment]
> I also found a problem with environment variables such as `os_getenvname_at_index` on Windows. @justinmk Are you planning on implementing #7920? I do not mind taking over #7920.

@erw7 I rebased #7920 and hope to merge it today. Is there still a problem for you after #7920?

**Comment by erw7** [comment]
> Is there still a problem for you after #7920?

Yes. #7920 can not solve multi byte environment variable problem. In order to solve it, I think https://github.com/neovim/neovim/pull/7920/commits/c6855dd85e8c35bcdab562d1a97e1c2f7f26ba69 is necessary.

**Comment by erw7** [comment]
> Is there still a problem for you after #7920?

Yes. A problem arises if multibyte is included in the environment variable passed to the child process(jobstart, terminal etc).

**Comment by erw7** [comment]
I think #8398 will not be solved in #7920. To solve this problem, I think https://github.com/neovim/neovim/pull/7920/commits/c6855dd85e8c35bcdab562d1a97e1c2f7f26ba69 is necessary.

**Comment by teto** [comment]
#7202 needs a rebase and there is a lingering crash. Feel free to takeover.

**Comment by justinmk** [comment]
> I think [c6855dd](https://github.com/neovim/neovim/commit/c6855dd85e8c35bcdab562d1a97e1c2f7f26ba69) is necessary.

But c6855dd assumes utf16, which isn't codepage specific. How can that work?

> Please do not merge #7920. It breaks the working `os_setenv`

Can you give a test case? Before #7920, the test added to `let_spec.lua` fails.

> I'm thinking about having libuv or neovim have utf-8 environment variable block.

@erw7  can you try #7920 ? I added `envmap` to env.c. Because the libuv 1.12 API (`uv_os_getenv`) returns allocated strings. It does UTF-8 conversion for us, but I'm not sure if it works for codepages besides UTF-16.

**Comment by erw7** [comment]
> But [c6855dd](https://github.com/neovim/neovim/commit/c6855dd85e8c35bcdab562d1a97e1c2f7f26ba69) assumes utf16, which isn't codepage specific. I'm surprised it would work.

In my understanding, Windows internally converts codepage and utf16, and has environment block of both utf16 and codepage.

See. [_environ, _wenviron](https://docs.microsoft.com/en-us/cpp/c-runtime-library/environ-wenviron?view=vs-2017)

> Can you give a test case?

To create it, I think we need to call `GetEnvironmentVariableW` using `ffi` on `let_spec.lua`. Please give me some time.

> @erw7 can you try #7920

#7920(https://github.com/neovim/neovim/pull/7920/commits/5c2180078a748e3c7faf19c260f876b8ce5e4e46) works perfectly.  However, [`os_getenvname_at_index`](https://github.com/neovim/neovim/blob/533d4a36ec03626e9d796ef7e2a9aa3c0e1ce7bf/src/nvim/os/env.c#L96) uses the environment variable block of the code page, so problems will occur with completing the multibyte environment variable name.

**Comment by erw7** [comment]
> Can you give a test case?

I tried to create it. How about with  https://github.com/erw7/neovim/commit/89bf5b1bc578dfa2338e4e8727d41ff2dc10be3a (https://github.com/neovim/neovim/commit/c6855dd85e8c35bcdab562d1a97e1c2f7f26ba69 failed this test). Is it also necessary for testing to pass multibyte environment variables from parent process to neovim(It was this test that needed `ffi`)?

**Comment by justinmk** [comment]
Thanks!

> Is it also necessary for testing to pass multibyte environment variables from parent process to neovim

I'm ok without it.

**Comment by justinmk** [comment]
@erw7 thanks for https://github.com/erw7/neovim/commit/89bf5b1bc578dfa2338e4e8727d41ff2dc10be3a , I included it in #7920  and will merge that now.

> However, [`os_getenvname_at_index`](https://github.com/neovim/neovim/blob/533d4a36ec03626e9d796ef7e2a9aa3c0e1ce7bf/src/nvim/os/env.c#L96) uses the environment variable block of the code page, so problems will occur with completing the multibyte environment variable name.

Tracked in https://github.com/neovim/neovim/issues/9655. I don't plan to work on this, so if you want to, please do.
